grafana/resource_alerting_contact_point.go

```diff
@@ -0,0 +1,449 @@
+	for _, uid := range uids {
+		if err := client.DeleteContactPoint(uid); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return diag.Diagnostics{}
+}
+
+func unpackContactPoints(data *schema.ResourceData) []gapi.ContactPoint {
+	result := make([]gapi.ContactPoint, 0)
+	name := data.Get("name").(string)
+	for _, n := range notifiers {
+		if points, ok := data.GetOk(n.meta().typeStr); ok {
+			for _, p := range points.([]interface{}) {
+				result = append(result, n.unpack(p, name))
+			}
+		}
+	}
+
+	return result
+}
+
+func packContactPoints(ps []gapi.ContactPoint, data *schema.ResourceData) {
+	pointsPerNotifier := map[notifier][]interface{}{}
+	for _, p := range ps {
+		data.Set("name", p.Name)
+
+		for _, n := range notifiers {
+			if p.Type == n.meta().typeStr {
+				packed := n.pack(p)
+				pointsPerNotifier[n] = append(pointsPerNotifier[n], packed)
+				continue
+			}
+		}
+	}
+
+	for n, pts := range pointsPerNotifier {
+		data.Set(n.meta().typeStr, pts)
+	}
+}
+
+func unpackCommonNotifierFields(raw map[string]interface{}) (string, bool, map[string]interface{}) {
+	return raw["uid"].(string), raw["disable_resolve_message"].(bool), raw["settings"].(map[string]interface{})
+}
+
+func packCommonNotifierFields(p *gapi.ContactPoint) map[string]interface{} {
+	return map[string]interface{}{
+		"uid":                     p.UID,
+		"disable_resolve_message": p.DisableResolveMessage,
+	}
+}
+
+func packSettings(p *gapi.ContactPoint) map[string]interface{} {
+	settings := map[string]interface{}{}
+	for k, v := range p.Settings {
+		settings[k] = fmt.Sprintf("%#v", v)
+	}
+	return settings
+}
+
+func commonNotifierResource() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"uid": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The UID of the contact point.",
+			},
+			"disable_resolve_message": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Whether to disable sending resolve messages.",
+			},
+			"settings": {
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Sensitive:   true,
+				Default:     map[string]interface{}{},
+				Description: "Additional custom properties to attach to the notifier.",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+const UIDSeparator = ";"
+
+func packUIDs(uids []string) string {
+	return strings.Join(uids, UIDSeparator)
+}
+
+func unpackUIDs(packed string) []string {
+	return strings.Split(packed, UIDSeparator)
+}
+
+func toUIDSet(uids []string) map[string]bool {
+	set := map[string]bool{}
+	for _, uid := range uids {
+		set[uid] = true
+	}
+	return set
+}
+
+type notifier interface {
+	meta() notifierMeta
+	schema() *schema.Resource
+	pack(p gapi.ContactPoint) interface{}
+	unpack(raw interface{}, name string) gapi.ContactPoint
+}
+
+type notifierMeta struct {
+	typeStr string
+	desc    string
+}
+
+type emailNotifier struct{}
```

Comment:
We might want to move the notifier implementations to another file, as it will eventually become super long and packed.